Depth sorting for node faces

x2048 opened the pull request:
This PR aims to implement correct painter's algorithm for transparent nodes on the map.

* Vertex indexes for all transparent faces are captured during mesh generation.
* [BSP Tree](https://en.wikipedia.org/wiki/Binary_space_partitioning) is generated for transparent triangles.
* When rendering the map, BSP tree is traversed to generate correct (in most cases) rendering order, which is applied by replacing indexes in the mesh buffers as suggested by @hecktest.

This PR supersedes parts of #11130 and fixes most of cases of transparency problems (including transparent nodebox nodes at the cost of performance).

## Handling Transparent Nodeboxes

In order to render overlapping nodeboxes correctly, this PR splits each nodebox at the edges of other nodeboxes along the same axis. This process produces a significantly larger number of 'atomic' nodeboxes that are guaranteed not to overlap, and may impact overall rendering performance. The total number of nodeboxes (defined + produced) is currently limited to 100.

## To do

This PR is Ready for Review and testing

## How to test

Use a combination of semi-transparent nodes such as:
* Alpha testing nodes in devtest
* Water
* Colored glass in Mineclone 2
* Stained glass 

Transparent nodes should be rendered correctly with minimal temporary artifacts from all angles.

x2048 committed:
Single meshbuffer transparency sorting

x2048 committed:
Allow multiple transparent materials per mapblock

Plus avoid sorting the transparent faces on every frame.

x2048 committed:
Disable tiling for semi-transparent materials

This is necessary to avoid depth sorting problems
caused by large triangles created by tiling.

Special case to disable clamp-to-edge for liquids is
necessary to render flowing liuids correctly.

This also fixes autogeneration of texture coordinates in
e.g. nodebox nodes to be compatible with clamp-to-edge.

x2048 committed:
Use BSP tree to order transparent triangles

https://en.wikipedia.org/wiki/Binary_space_partitioning

x2048 committed:
Move BSP tree creation to mesh generation thread

x2048 committed:
Minor correction

x2048 committed:
Fix rendering of complex transparent nodeboxes

1. Transparent nodebox node is now split correctly into
non-intersecting boxes/triangles
2. BSP algorithm now uses largest triangle as
the split plane when looking inside of a node.

x2048 commented:
Tagging @appgurueu @Wuzzy2 @sfan5 @Andrey2470T @hecktest

x2048 committed:
Re-enable 2s rendering time guard

x2048 committed:
Fix typo

Andrey2470T commented:
Seems it has fixed the main transparency issue, transparent nodes are rendered behind other transparent ones now... but:
![screenshot_20211016_235712](https://user-images.githubusercontent.com/25750346/137601962-6aa86c81-cdd8-4f45-8ab5-1ad07e9f4683.png)

I used the colored glass from `glass_stained` mod and it caused that artifact, some parts of mesh faces of the nodes are black and those, as I understood, that are beyond the node area. Here's how it looks like in minetest without the bugfix:
![screenshot_20211016_235641](https://user-images.githubusercontent.com/25750346/137602098-95625723-647e-4bbc-81a6-a21734e7e915.png)

x2048 committed:
Remove changes to texture tiling

Use more precise and compatible way of avoiding merge of
transparent face tiles via TileSpac.isTileable()

x2048 committed:
Fix depth sorting for oversized node boxes.

Node example: glass_stained:glass_1_noncuple

x2048 committed:
Fix compatibility with shadow mapping

x2048 committed:
Skip transparency sorting for distant map blocks.

Improve rendering performance by reducing the number of drawcalls / setMaterial() calls.

Andrey2470T commented:
Now after the fix FPS has become higher on that server, even a bit higher 8 (varies within 10 - 13 FPS). Thanks for the fix.

Andrey2470T commented:
I've discovered a wonderful bug with this PR (server `Tunnelers Abyss`):
![Снимок экрана от 2021-11-04 23-34-20](https://user-images.githubusercontent.com/25750346/140417125-c1b9cd67-da8f-4f6e-9930-979904d5bbd8.png)

![Снимок экрана от 2021-11-04 23-37-10](https://user-images.githubusercontent.com/25750346/140417196-130dc47b-4f3d-4736-9c74-fbe362e2506c.png)

![Снимок экрана от 2021-11-04 23-37-31](https://user-images.githubusercontent.com/25750346/140417243-5a7167fc-03eb-486f-b24a-7c1d48ba06c7.png)

Looks like as if somebody strongly extruded the colored glass panes to the outside.

Andrey2470T commented:
So is this bug irrelevant to this PR if you told me you didn't discover this on Tunneler's Abyss? I don't see its display on any another builds, but only in your one. I doubt this problem concerns to my hardware or video drivers. So, I can conclude this bug hides somewhere in your PR's code.

x2048 commented:
I can confirm the bug, but it does not occur in 100% of cases.

x2048 committed:
Improve splitting of transparent nodeboxes

* Tie default splits to block size (BS)
* Round edges of defined boxes to 1E-3
* Limit the total number of boxes to 100
* Avoid uninitialized reads

x2048 commented:
@Andrey2470T The bug was caused by an uninitialized memory read due to reallocation inside std::vector and the reference to the current box not being adjusted. It should be fixed now, and I've also added a few improvements to the code to avoid performance problems in edge cases.

appgurueu commented:
Could the BSP tree be made to take into account semitransparent particle & entity triangles?

x2048 commented:
I would not say it is impossible, but it is a significant step up in complexity. There are multiple reasons for this:
* Current BSP tree is per mapblock and is built on assumptions of a mapblock mesh (e.g. axis-aligned cubes). Building a generic BSP tree for free-form objects requires some good math and mesh manipulation on the fly (adding new vertices).
* BSP trees in general are expensive to build and are usually used for static meshes (like the map blocks). Putting a moving object into a BSP tree means rebuilding the tree on each frame, and it's at least O(N logN) on the number of triangles.
* Our current rendering process simply not fit for this at the moment, we would need to extract a deferred transparent rendering pass, and I am not sure Irrlicht is helpful here.

Andrey2470T commented:
Is this PR ready for merging? In any case, I didn't found any other noticable bugs, except which I reported about before and which already fixed. The only current downside is a lower FPS, however generally it is not critical and quite playable.

x2048 commented:
This PR is ready to be reviewed by the core developers, there is no known technical debt / problems to be fixed.

erlehmann commented:
@x2048 please rebase this branch on master so it can be compared to recent Minetest easier in terms of performance.

erlehmann commented on src/client/clientmap.cpp:
```diff
@@ -389,35 +414,14 @@ void ClientMap::renderMap(video::IVideoDriver* driver, s32 pass)
 					video::SMaterial& material = buf->getMaterial();
 					video::IMaterialRenderer* rnd =
-						driver->getMaterialRenderer(material.MaterialType);
+							driver->getMaterialRenderer(material.MaterialType);
```
Is the added whitespace here really necessary?

erlehmann commented on src/client/clientmap.cpp:
```diff
@@ -742,49 +776,68 @@ void ClientMap::renderMapShadows(video::IVideoDriver *driver,
 
-			drawcall_count += list.bufs.size();
+		// Check and abort if the machine is swapping a lot
+		if (draw.getTimerTime() > 1000) {
```
What kind of scenario triggers this?

x2048 replied:
Looks like going under roughly 1fps will trigger this condition.

runsy commented:
Yes. I have issues:

[](https://forum.minetest.net/download/file.php?id=25605)

Transparent nodes besides othe rtransparent nodes.

x2048 commented:
@runsy Can you give more detail on how to reproduce it?

wsor4035 commented:
@x2048 seems it was taken from https://forum.minetest.net/viewtopic.php?p=406861#p406861 which appears to be regular mte and not this PR? follow up https://forum.minetest.net/viewtopic.php?p=407177#p407177

x2048 commented:
@runsy The problem on your screenshot is fixed by this PR.

@wsor4035 Thank you for the pointers.

runsy commented:
> @x2048 seems it was taken from https://forum.minetest.net/viewtopic.php?p=406861#p406861 which appears to be regular mte and not this PR? follow up https://forum.minetest.net/viewtopic.php?p=407177#p407177

I expect my issue solved with this PR. I do not check it.

Ca n anyone provide me this PR in ZIP to compile it in order to test it?

runsy commented:
> @runsy The problem on your screenshot is fixed by this PR.
> 
> @wsor4035 Thank you for the pointers.

Cool. So when merged? I expect this in 5.6.

sfan5 commented on src/client/clientmap.cpp:
```diff
@@ -100,6 +100,28 @@ ClientMap::ClientMap(
 
 }
 
+void ClientMap::updateCamera(const v3f &pos, const v3f &dir, f32 fov, const v3s16 &offset)
```
general advice: vectors are only a couple bytes so using a constref for them doesn't make much sense

x2048 replied:
I was hoping that it would allow passing arguments in registers, rather than pushing on stack.

sfan5 replied:
While a pointer definitely fits in a register I think having to dereference it is still worse than reading everything close together from the stack.
Anyway, there are still a few more `const v3s16 &` and `const v3f &`

x2048 replied:
Done

sfan5 commented on src/client/clientmap.cpp:
```diff
@@ -323,22 +345,17 @@ void ClientMap::renderMap(video::IVideoDriver* driver, s32 pass)
+		Update transparent meshes
+	*/
+	if (is_transparent_pass)
+		this->updateTransparentMeshBuffers();
```
```suggestion
		updateTransparentMeshBuffers();
```

sfan5 commented on src/client/clientmap.cpp:
```diff
@@ -742,49 +776,68 @@ void ClientMap::renderMapShadows(video::IVideoDriver *driver,
+			// iterate in reverse to draw closest blocks first
+			for (auto it = list.bufs.rbegin(); it != list.bufs.rend(); ++it) {
+				draw_order.emplace_back(it->first, it->second, it != list.bufs.rbegin());
+			}
```
Could use a reserve call *before* the loop, this probably works:
`draw_order.reserve(draw_order.size() + list.bufs.size());`

x2048 replied:
I decided to preallocate draw_order once, since there are a lot of buffers, and moving memory feels more expensive.

sfan5 commented on src/client/clientmap.cpp:
```diff
@@ -891,3 +945,39 @@ void ClientMap::updateDrawListShadow(const v3f &shadow_light_pos, const v3f &sha
+	ScopeProfiler sp(g_profiler, "CM::updateTransparentMeshBuffers", SPT_AVG);
+	u32 sorted_blocks = 0;
+	u32 unsorted_blocks = 0;
+	const f32 MAX_SORT_DISTANCE = BS * BS * 75 * 75; // 75 nodes, 1.5x animation threshold
```
this might be clearer written as `powf(75 * BS, 2)`

sfan5 commented on src/client/content_mapblock.cpp:
```diff
@@ -381,12 +381,12 @@ void MapblockMeshGenerator::drawAutoLightedCuboid(aabb3f box, const f32 *txc,
 		box.MaxEdge *= f->visual_scale;
 	}
-	box.MinEdge += origin;
-	box.MaxEdge += origin;
```
If this introduces a new requirement that origin must be added *after* calling `drawNodeboxNode` that should be written down where `drawNodeboxNode` is defined or so.

x2048 replied:
Origin is added in line 388 below, I just moved generation of texture coords earlier, because having origin added breaks it.

x2048 replied:
It does not. I merely moved generation of texture coordinates earlier because it was broken by adding origin too early.

sfan5 commented on src/client/content_mapblock.cpp:
```diff
@@ -1377,6 +1377,47 @@ void MapblockMeshGenerator::drawNodeboxNode()
+			}
+
+			// split the boxes at recorded sections
+			for (size_t i = 0; i < boxes.size() && i < 100; i++) {
```
why stop at 100?

x2048 replied:
This is a silly cap to prevent crashing away in case the inner loop starts adding an infinite number of new boxes due to e.g. precision glitches.

Number 100 is an arbitrary, reasonably high number that seems to work with the nodes defined today.

sfan5 replied:
That's okay, just please write it down in a comment.

x2048 replied:
done

sfan5 commented on src/client/content_mapblock.cpp:
```diff
@@ -1377,6 +1377,47 @@ void MapblockMeshGenerator::drawNodeboxNode()
+	}
+
+	if (isTransparent) {
+		for (int axis = 0; axis < 3; axis++) {
+			// identify sections
+			std::vector<float> sections;
```
this can be optimized regarding allocations like so:
```suggestion
		std::vector<float> sections;
		sections.reserve(8 + 2 * boxes.size());
		for (int axis = 0; axis < 3; axis++) {
			// identify sections
			sections.clear();
```

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1003,6 +1004,176 @@ static void applyTileColor(PreMeshBuffer &pmb)
+s32 MapBlockBspTree::buildTree(v3f normal, v3f origin, float delta, const std::vector<s32> &list, u32 depth)
+{
+	// if the list is empty, don't bother
+	if (list.size() == 0) {
```
generally you can replace all `.size() == 0` or `.size() > 0` here with `.empty()` or `!.empty()` respectively

x2048 replied:
replaced

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1003,6 +1004,176 @@ static void applyTileColor(PreMeshBuffer &pmb)
+
+	nodes.clear();
+
+	std::vector<s32> indexes;
```
```suggestion
	std::vector<s32> indexes;
	indexes.reserve(triangles->size());
```

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1003,6 +1004,176 @@ static void applyTileColor(PreMeshBuffer &pmb)
+
+	// if there is only one triangle, or the delta is insanely small, this is a leaf node
+	if (list.size() == 1 || delta < 0.01) {
+		struct TreeNode node;
+		node.normal = normal;
+		node.origin = origin;
+		node.triangle_refs = list;
+		node.front_ref = -1;
+		node.back_ref = -1;
```
create a constructor for this?

x2048 replied:
Done

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1173,8 +1344,31 @@ MapBlockMesh::MapBlockMesh(MeshMakeData *data, v3s16 camera_offset):
-			buf->append(&p.vertices[0], p.vertices.size(),
-				&p.indices[0], p.indices.size());
+			switch (p.layer.material_type) {
+			// list of transparent materials taken from tile.h
```
wasn't `isTransparent()` intended for this?

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1298,6 +1493,68 @@ bool MapBlockMesh::animate(bool faraway, float time, int crack,
+	for (auto i : triangle_refs) {
+		const auto &t = m_transparent_triangles[i];
+		if (current_buffer != t.buffer) {
+			if (current_buffer != nullptr) {
+				this->m_transparent_buffers.emplace_back(current_buffer, current_strain);
```
```suggestion
			if (current_buffer) {
				m_transparent_buffers.emplace_back(current_buffer, current_strain);
```

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1298,6 +1493,68 @@ bool MapBlockMesh::animate(bool faraway, float time, int crack,
+		current_strain.push_back(t.p3);
+	}
+
+	if (current_strain.size() > 0) {
+		this->m_transparent_buffers.emplace_back(current_buffer, current_strain);
+	}
```
```suggestion
	if (!current_strain.empty())
		m_transparent_buffers.emplace_back(current_buffer, current_strain);
```

sfan5 commented on src/client/mapblock_mesh.h:
```diff
@@ -71,6 +71,84 @@ struct MeshMakeData
+	scene::SMeshBuffer *m_buffer;
+	std::vector<u16> m_vertex_indexes;
+
+	friend class MapBlockMesh;
```
I don't remember seeing any code that'd require this.

x2048 replied:
Unfriended

sfan5 commented on src/client/mapblock_mesh.h:
```diff
@@ -125,6 +203,15 @@ class MapBlockMesh
+	void consolidateTransparentBuffers();
+
+	/// get the list of transparent buffers
+	std::vector<PartialMeshBuffer> &getTransparentBuffers()
```
could be const

x2048 replied:
It's now const along with a bunch of other methods.

sfan5 commented on src/nodedef.cpp:
```diff
@@ -666,7 +666,7 @@ void ContentFeatures::deSerialize(std::istream &is)
 		const TileSpec &tile, const TileDef &tiledef, video::SColor color,
 		u8 material_type, u32 shader_id, bool backface_culling,
-		const TextureSettings &tsettings)
+		const TextureSettings &tsettings, AlphaMode alpha)
```
looks unused

x2048 replied:
Removed

sfan5 commented on src/client/tile.h:
```diff
@@ -259,6 +259,31 @@ struct TileLayer
 			&& (material_flags & MATERIAL_FLAG_TILEABLE_VERTICAL);
 	}
 
+	bool isLiquid() const
```
looks unused

x2048 replied:
Removed

runsy commented:
@sfan5 Please review this one.

sfan5 reviewed (changes requested):
This review was mostly reading code though I gave it a try ingame and tested a few random things, it was fine.

sfan5 commented:
Question: Is this whole ordeal necessary for clipped alpha (no blending)? The code seems to do it for those (or at least `isTransparent` checks them too...).

appgurueu commented:
> Question: Is this whole ordeal necessary for clipped alpha (no blending)? The code seems to do it for those (or at least `isTransparent` checks them too...).

Not if we have a `discard` in our shaders.

x2048 committed:
Style and performance improvements

Co-authored-by: sfan5 <sfan5@live.de>

lhofhansl commented:
Part of this is already merged. Do we still need this one? @x2048

x2048 commented:
@lhofhansl Yes we do. We merge per-mapblock sorting, this adds per-face sorting.

kilbith commented:
@x2048 [PTAL](https://irc.minetest.net/minetest-dev/2022-02-23#i_5943075).

x2048 commented:
@kilbith OIT is a game changer.
@appgurueu  too bad I did not see https://jcgt.org/published/0002/02/09/ earlier.

@sfan5 @lhofhansl would you still like to see this merged, until order-independent transparency is implemented?

x2048 committed:
Clean up code

* Remove TileLayer::isLiquid()
* Remove alpha parameter from fillTileAttribs()
* Add const qualifier to MapBlockMesh::getTransparentBuffers()

Remove friend MapBlockMesh from PartialMeshBuffer

x2048 committed:
Preserve vertex buffer when pathing mesh buffer indexes

x2048 committed:
Preallocate draw_order when rendering ClientMap

x2048 committed:
Clean up code #2

* Make use of std::vector::empty()
* Annotate the limit of 100 nodebox splits
* Pass v3s16 by value, not reference

SmallJoker commented:
What's the best approach to compare the performance of this PR vs master? Is it sufficient to compare the `renderMap` F6 profiler entries? Do I need mapblocks filled with transparent nodes (glass?) to trigger the new code?

> would you still like to see this merged, until order-independent transparency is implemented?

What would your plans be in this case? I cannot tell what would be better in this case due to lack of knowledge in the rendering code, but having a general approach for all rendered parts (objects, nodes) would be really compelling to keep a simple/clear code structure.

x2048 committed:
Fix compilation on Win32 using proper type

sfan5 commented:
> @sfan5 @lhofhansl would you still like to see this merged, until order-independent transparency is implemented?

sure

x2048 commented:
@SmallJoker I created a pattern out of transparent nodes in devtest to trigger the worst-case scenario:
![image](https://user-images.githubusercontent.com/4933697/156660103-3f8c39dd-5c65-48f4-a645-c474ec5c2afa.png)
The height of each column is 44 nodes.

This case renders at 60 FPS without depth sorting and at 10-12 FPS with depth sorting on my laptop (Ryzen 7), so there is definitely a performance penalty associated.

runsy commented:
> @SmallJoker I created a pattern out of transparent nodes in devtest to trigger the worst-case scenario: ![image](https://user-images.githubusercontent.com/4933697/156660103-3f8c39dd-5c65-48f4-a645-c474ec5c2afa.png) The height of each column is 44 nodes.
> 
> This case renders at 60 FPS without depth sorting and at 10-12 FPS with depth sorting on my laptop (Ryzen 7), so there is definitely a performance penalty associated.

Too much FPS drop. HUGE. REALLY ANNOYING.

Any better solution?

My ideas:

1. Only transparency til a depth (maybe 3 - 5 nodes)
2. An optional feature. Disabled by default.
3. A better algorithm. Study the Minecraft / other games approach.

x2048 commented:
@runsy Does it mean that you suggest closing this PR without merging it in?

I explicitly showed the worst case with 22x22x44 nodes with 4 materials and lots of overlaps. There is a limit of 75 nodes for transparency sorting, about five blocks from the player, and I can of course make it configurable.

OIT has a good chance to be faster but will increase average load on the GPU, so not everyone will benefit from it. This is a topic for a different PR.

Wuzzy2 commented:
How about adding a setting which sets a maximum as to how much complexity is allowed, to allow fine-tuning if neccessary. That way, players with weak computers can turn it off or set to a low value.

runsy commented:
> @runsy Does it mean that you suggest closing this PR without merging it in?

What I'm saying is that since it makes frame drops of 80% with respect to normal, make it optional. Or as sfan5 says tuneable.

appgurueu commented:
> > @runsy Does it mean that you suggest closing this PR without merging it in?
> 
> What I'm saying is that since it makes frame drops of 80% with respect to normal (60FPS -> 10 FPS is too much, a frame drop of 5-10FPS would be reasonable), make it optional. Or as wuzzy says tuneable.

in the *worst case*

runsy commented:
> > > @runsy Does it mean that you suggest closing this PR without merging it in?
> > 
> > 
> > What I'm saying is that since it makes frame drops of 80% with respect to normal (60FPS -> 10 FPS is too much, a frame drop of 5-10FPS would be reasonable), make it optional. Or as wuzzy says tuneable.
> 
> in the _worst case_

I always heard that minetets supported old equipment and blah blah blah. Let's keep it real.

appgurueu commented:
> > in the _worst case_
> 
> I always heard that minetets supported old equipment and blah blah blah. Let's keep it real.

You're the one not keeping it real by arguing using the worst case (the only valid point I could see is griefers abusing this to lower FPS of other players in certain areas). If semitransparent nodes are randomly distributed, this worst case definitely won't come into play.

To give another example: Quicksort has a worst-case complexity of O(n²) as well, yet it is widely used, because the expected average running time if randomly picking pivots is O(n log n). Granted, in this case the complexity isn't determined by the *data* (the list) being "average", but rather the random not constantly picking the worst pivots.

runsy commented:
So, what is the final decision?

It is necessary another dev to review this.

x2048 commented:
@runsy The distance of depth sorting will become adjustable, which will allow the client to avoid performance drops. The PR goes in, and we start work on OIT that would support both map, entities, clouds and particles. And be less sensitive to edge cases.

runsy commented:
> @runsy The distance of depth sorting will become adjustable, which will allow the client to avoid performance drops. The PR goes in, and we start work on OIT that would support both map, entities, clouds and particles. And be less sensitive to edge cases.

coolz

appgurueu commented:
> coolz

Please use a reaction instead of a message for this. Related: Could we use some guidelines for discussion on GH?

sfan5 commented:
@x2048 if you want this merged there's only two open comments from my side.

SmallJoker commented:
![grafik](https://user-images.githubusercontent.com/1497498/159158290-328e6aef-f2e4-4e22-91f4-6c077e88e1a3.png)

Basically instant OOM while facing transparent nodeboxes near the spawn of Blocky Survival (mt.krypticbit.com 30005). This definitely need investigation.

x2048 commented:
@SmallJoker good stuff, thanks for sharing

x2048 committed:
Fix runaway loop when building BSP tree.

Due to index variable being u16 loop would lead to OOM and crash
when a mapblock has more than 64K triangles.

x2048 commented:
@SmallJoker OOM should be fixed now.

x2048 committed:
Add setting to control transparency sorting distance

Setting name: transparency_sorting_distance

x2048 commented:
@runsy i've now added a setting `transparency_sorting_distance` to control the performance impact of transparency sorting.

x2048 committed:
Add constructor to MapBlockBspTree::MeshTriangle

runsy commented:
> @runsy i've now added a setting `transparency_sorting_distance` to control the performance impact of transparency sorting.

Coolz Thankz.

x2048 committed:
Pass primitive vector types by value.

Seems to be faster and smaller code than passing by const-reference

SmallJoker commented on src/client/content_mapblock.cpp:
```diff
@@ -1377,6 +1377,52 @@ void MapblockMeshGenerator::drawNodeboxNode()
+		std::vector<float> sections;
+		sections.reserve(8 + 2 * boxes.size());
+		for (int axis = 0; axis < 3; axis++) {
+			// identify sections
+			sections.clear();
+
+			// Default split at node bounds, up to 3 nodes in each direction
+			for (float s = -3.5f * BS; s < 4.0f * BS; s += 1.0f * BS)
+				sections.push_back(s);
```
Might it be more efficient if you preload `sections` with these values outside of the loop and use `sections.resize(n);`?
However I do not know whether this affects the capacity changed by `sections.reserve` above.

x2048 replied:
done

SmallJoker commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1003,6 +1004,164 @@ static void applyTileColor(PreMeshBuffer &pmb)
+	root = buildTree(v3f(1, 0, 0), v3f(85, 85, 85), 40, indexes, 0);
+}
+
+static const MeshTriangle *findSplitCandidate(const std::vector<s32> &list, const std::vector<MeshTriangle> &triangles)
```
Would you please be so nice to add a description to this function? It would be nice to at least get an idea of how this function helps the depth sorting. More helpful comments = easier to understand for maintenance.

x2048 replied:
Done for this function and added a link to the method to BspTree class.

SmallJoker commented on src/client/mapblock_mesh.cpp:
```diff
@@ -1003,6 +1004,164 @@ static void applyTileColor(PreMeshBuffer &pmb)
+
+	// find the triangle with the largest area and closest to the center
+	const MeshTriangle *candidate_triangle = &triangles[list[0]];
+	const MeshTriangle *ith_triangle;
+	for (s32 i : list) {
+		ith_triangle = &triangles[i];
```
```C++
for (const MeshTriangle &t : list) {
	if (t.areaSQ > candidate_triangle->areaSQ ||
```
The range-based `for` loop is shorter, supersedes an additional variable and `t` still allows address-of operators for the return value.

x2048 replied:
Well, `list` contains indexes of the triangles, and I get the actual triangle from `triangles`. I don't think I an do this in a different way.

SmallJoker replied:
Oh yes, you are totally right. :+1:

SmallJoker reviewed (approved):
Works. Code adjustments are welcome.

x2048 committed:
Improvements suggested by @SmallJoker

sfan5 reviewed (approved):
Gave it a quick test, I don't have any intensive test scenes so I trust @SmallJoker with that.
